[symm_mem] Add nccl as a backend for symmetric memory

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #155740

Running unit test:

 TORCH_SYMMMEM=NCCL TORCH_DISTRIBUTED_DEBUG=INFO TORCH_CPP_LOG_LEVEL=INFO pytest test/distributed/test_nccl.py -k test_nccl_symmem_alloc

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

fduwjj committed:
[symm_mem] Add nccl as a backend for symmetric memory

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -0,0 +1,352 @@
+      const std::string& group_name,
+      ncclWindow_t handle,
+      ncclWindow_t signal_handle)
+      : allocation_(allocation),
```
```suggestion
      : allocation_(std::move(allocation)),

Skylion007 commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -0,0 +1,352 @@
+    // WIP
+  }
+
+  ~NCCLSymmetricMemory() override{
```
```suggestion
  ~NCCLSymmetricMemory() override = default;

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -0,0 +1,352 @@
+    TORCH_CHECK(ncclPg != nullptr, "backend must be a NCCL process group");
+    ncclComm_t comm = reinterpret_cast<ncclComm_t>(ncclPg->getCommPtr());
+    C10D_NCCL_CHECK(
+      ncclCommWindowRegister(comm, ptr, it->second->buffer_size, (ncclWindow_t*)&handle, NCCL_WIN_COLL_SYMMETRIC),
```
There are multiple flags possible:
https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/flags.html#win-flags

fduwjj replied:
> Register buffer into NCCL window, and users need to guarantee the offset to the buffer head address from all ranks must be equal when calling NCCL collective operations. It allows NCCL to operate buffer in a symmetric way and provide the best performance.

So I guess for the perf reason we should use this one?

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -0,0 +1,352 @@
+
+    void* signal_pad_ptr;
+    TORCH_CHECK(ncclMemAlloc(&signal_pad_ptr, signal_pad_size) == ncclSuccess, "ncclMemAlloc failed");
+    C10D_NCCL_CHECK(
+    ncclCommWindowRegister(comm, signal_pad_ptr, signal_pad_size, (ncclWindow_t*)&signal_handle, NCCL_WIN_COLL_SYMMETRIC),
+    c10::str(
```
lint

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -0,0 +1,352 @@
+#ifdef USE_C10D_NCCL
+
+#if NCCL_VERSION_CODE >= NCCL_VERSION(2, 27, 1)
+#define NCCL_HAS_SYMMEM_SUPPORT
+#endif
```
You would need `include <nccl.h>` at least?

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -0,0 +1,352 @@
+
+
+    auto group = resolve_process_group(group_name.value());
+    c10::cuda::CUDAGuard guard(it->second->device_idx);
```
nit: give a reference name to `it->second` so that the code becomes cleaner down the line?

kwen2501 reviewed (commented):
Does it compile? Any change needed in CMakeList? 
Otherwise, LGTM.

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -0,0 +1,352 @@
+    c10::cuda::CUDAGuard guard(device_idx);
+    // TODO: we might need to use a roundup or mempool for mem allocation.
+    void* ptr;
+    TORCH_CHECK(ncclMemAlloc(&ptr, size) == ncclSuccess, "ncclMemAlloc failed");
```
nit: use `NCCLCHECK` instead?

fduwjj committed:
Update on "[symm_mem][WIP] Add nccl as a backend for symmetric memory"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[symm_mem][WIP] Add nccl as a backend for symmetric memory"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[symm_mem][WIP] Add nccl as a backend for symmetric memory"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
Somehow I saw this error logs when running unit test

[rank0]:[I617 21:49:04.240959859 ProcessGroupNCCL.cpp:1479] [PG ID 0 PG GUID 0(default_pg) Rank 0] Watchdog joined, destroying NCCL communicators.
[rank3]:[I617 21:49:04.240961582 ProcessGroupNCCL.cpp:1479] [PG ID 0 PG GUID 0(default_pg) Rank 3] Watchdog joined, destroying NCCL communicators.

[2025-06-17 21:49:04] devvm5552:3355146:3355146 [0] transport/p2p.cc:1128 NCCL WARN Cuda failure 1 'invalid argument'

[2025-06-17 21:49:04] devvm5552:3355146:3355146 [0] init.cc:2143 NCCL WARN commReclaim: cleanup comm 0x9801e10 rank 0 failed in destroy/abort, error 1
[rank0]:[I617 21:49:04.735740424 ProcessGroupNCCL.cpp:1487] [PG ID 0 PG GUID 0(default_pg) Rank 0] Destroy complete.

[2025-06-17 21:49:04] devvm5552:3355147:3355147 [1] transport/p2p.cc:1128 NCCL WARN Cuda failure 1 'invalid argument'

[2025-06-17 21:49:04] devvm5552:3355147:3355147 [1] init.cc:2143 NCCL WARN commReclaim: cleanup comm 0xa6e0f00 rank 1 failed in destroy/abort, error 1
[rank1]:[I617 21:49:04.752210537 ProcessGroupNCCL.cpp:1487] [PG ID 0 PG GUID 0(default_pg) Rank 1] Destroy complete.

[2025-06-17 21:49:04] devvm5552:3355149:3355149 [2] transport/p2p.cc:1128 NCCL WARN Cuda failure 1 'invalid argument'

[2025-06-17 21:49:04] devvm5552:3355149:3355149 [2] init.cc:2143 NCCL WARN commReclaim: cleanup comm 0xa6e0de0 rank 2 failed in destroy/abort, error 1
[rank2]:[I617 21:49:04.800820235 ProcessGroupNCCL.cpp:1487] [PG ID 0 PG GUID 0(default_pg) Rank 2] Destroy complete.

[2025-06-17 21:49:04] devvm5552:3355150:3355150 [3] transport/p2p.cc:1128 NCCL WARN Cuda failure 1 'invalid argument'

[2025-06-17 21:49:04] devvm5552:3355150:3355150 [3] init.cc:2143 NCCL WARN commReclaim: cleanup comm 0xa6e17d0 rank 3 failed in destroy/abort, error 1

is this expected?

fduwjj commented on test/distributed/test_nccl.py:
```diff
@@ -239,6 +242,45 @@ def test_reduce_scatter(self, device, dtype):
+    @skip_but_pass_in_sandcastle_if(TEST_WITH_ROCM, "Skip NCCL tests for ROCm")
+    @skip_but_pass_in_sandcastle_if(IS_WINDOWS, "NCCL doesn't support Windows")
+    def test_nccl_symmem_alloc(self):
+        default_symm_mem_backend = os.getenv("TORCH_SYMMMEM")
+        os.environ["TORCH_SYMMMEM"] = "NCCL"
```
yes I will remove them.

fduwjj committed:
Update on "[symm_mem] Add nccl as a backend for symmetric memory"


Running unit test:

 TORCH_SYMMMEM=NCCL TORCH_DISTRIBUTED_DEBUG=INFO TORCH_CPP_LOG_LEVEL=INFO pytest test/distributed/test_nccl.py -k test_nccl_symmem_alloc

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj committed:
Update

[ghstack-poisoned]

fduwjj commented:
After rebase different tests failed, so the previous dynamo one is not related. I will merge this PR.

fduwjj committed:
Update on "[symm_mem] Add nccl as a backend for symmetric memory"


Running unit test:

 TORCH_SYMMMEM=NCCL TORCH_DISTRIBUTED_DEBUG=INFO TORCH_CPP_LOG_LEVEL=INFO pytest test/distributed/test_nccl.py -k test_nccl_symmem_alloc

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge -i

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -0,0 +1,355 @@
+          comm));
+
+    void* signal_pad_ptr;
+    TORCH_CHECK(ncclMemAlloc(&signal_pad_ptr, signal_pad_size) == ncclSuccess, "ncclMemAlloc failed");
```
tbh I don't know about this part, I just follow the implementation of cuda.
